File: frontend/src/domains/Citizen/features/Traffic/Models/TrafficIncident.js
// Domain Model

/**
 * TrafficIncident Domain Model
 * Follows OOP principles and encapsulates incident data
 * Immutable and self-validating
 */

/**
 * TrafficIncident Domain Model
 * ✅ UPDATED: Added `type` field to match backend
 * Backend sends: { id, title, lat, lng, level, type, image, ... }
 */

import {TRAFFIC_LEVELS, VALIDATION_RULES} from "../Config/trafficConfig.js";

export class TrafficIncident {
    constructor({
                    id,
                    title,
                    description,
                    lat,
                    lng,
                    level,
                    type,           //update : NEW: CAR,BIKe,ACCident,JAM
                    image,
                    timestamp,
                    reporter,
                    status = "PENDING",
                }) {
        // Validate on construction
        this._validate({id, title, lat, lng, level,type});

        // Immutable properties
        this._id = id;
        this._title = title;
        this._description = description || "";
        this._lat = lat;
        this._lng = lng;
        this._level = level;
        this._type = type || "ACCIDENT"; // ✅ Default to ACCIDENT
        this._image = image || null;
        this._timestamp = timestamp || Date.now();
        this._reporter = reporter || "Anonymous";
        this._status = status;

        Object.freeze(this); // Make immutable
    }

    // ==================== GETTERS ====================
    get id() {
        return this._id;
    }

    get title() {
        return this._title;
    }

    get description() {
        return this._description;
    }

    get lat() {
        return this._lat;
    }

    get lng() {
        return this._lng;
    }

    get level() {
        return this._level;
    }

    get type() {
        return this._type; //  NEW
    }

    get image() {
        return this._image;
    }

    get timestamp() {
        return this._timestamp;
    }

    get reporter() {
        return this._reporter;
    }

    get status() {
        return this._status;
    }

    get coords() {
        return [this._lng, this._lat];
    }

    // ==================== BUSINESS LOGIC ====================
    isHighPriority() {
        return this._level === TRAFFIC_LEVELS.HIGH;
    }

    isMediumPriority() {
        return this._level === TRAFFIC_LEVELS.MEDIUM;
    }

    isLowPriority() {
        return this._level === TRAFFIC_LEVELS.LOW;
    }

    isValidated() {
        return this._status === "VALIDATED";
    }

    isPending() {
        return this._status === "PENDING";
    }

    hasImage() {
        return this._image !== null && this._image !== undefined;
    }

    //  NEW: Check incident type
    isAccident() {
        return this._type === "ACCIDENT";
    }

    isTrafficJam() {
        return this._type === "JAM";
    }

    // ==================== VALIDATION ====================
    _validate({id, title, lat, lng, level,type}) {
        const errors = [];

        if (!id) {
            errors.push("ID is required");
        }

        if (!title || title.trim().length === 0) {
            errors.push("Title is required");
        }

        if (!Number.isFinite(lat) || lat < -90 || lat > 90) {
            errors.push("Invalid latitude");
        }

        if (!Number.isFinite(lng) || lng < -180 || lng > 180) {
            errors.push("Invalid longitude");
        }

        if (!VALIDATION_RULES.VALID_LEVELS.includes(level)) {
            errors.push(`Invalid level. Must be one of: ${VALIDATION_RULES.VALID_LEVELS.join(", ")}`);
        }

        // ✅ NEW: Validate type
        const validTypes = ["CAR", "BIKE", "ACCIDENT", "JAM", "SLOW", "FAST"];
        if (type && !validTypes.includes(type)) {
            errors.push(`Invalid type. Must be one of: ${validTypes.join(", ")}`);
        }

        if (errors.length > 0) {
            throw new Error(`TrafficIncident validation failed: ${errors.join(", ")}`);
        }
    }

    // ==================== UTILITIES ====================
    toJSON() {
        return {
            id: this._id,
            title: this._title,
            description: this._description,
            lat: this._lat,
            lng: this._lng,
            level: this._level,
            type: this._type,     //  NEW
            image: this._image,
            timestamp: this._timestamp,
            reporter: this._reporter,
            status: this._status,
        };
    }

    toString() {
        return `TrafficIncident(${this._id}: ${this._title} at [${this._lat}, ${this._lng}])`;
    }

    // ==================== FACTORY METHODS ====================
    static fromAPI(apiData) {
        return new TrafficIncident({
            id: apiData.id,
            title: apiData.title,
            description: apiData.description,
            lat: apiData.lat,
            lng: apiData.lng,
            level: apiData.level,
            type: apiData.type,   // ✅ NEW: Backend sends this
            image: apiData.image,
            timestamp: apiData.timestamp,
            reporter: apiData.reporter,
            status: apiData.status || "VALIDATED",
        });
    }

    static fromMock(mockData) {
        return new TrafficIncident({
            id: mockData.id,
            title: mockData.title,
            description: mockData.description,
            lat: mockData.lat,
            lng: mockData.lng,
            level: mockData.level,
            type: mockData.type || "ACCIDENT", // ✅ NEW
            image: mockData.image,
            timestamp: mockData.timestamp || Date.now(),
            reporter: mockData.reporter || "Mock User",
            status: "VALIDATED",
        });
    }
}

File: backend/src/main/java/com/urbanmonitor/domain/citizen/entity/TrafficIncident.java
package com.urbanmonitor.domain.citizen.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

/**
 * TrafficIncident Entity
 * Represents validated traffic incidents displayed on the map
 * 
 * Design Patterns:
 * - Builder Pattern (via @Builder)
 * - Entity Pattern (JPA)
 * 
 * OOP Principles:
 * - Encapsulation (private fields with getters/setters)
 * - Data validation via Bean Validation
 */
@Entity
@Table(name = "traffic_incidents", indexes = {
    @Index(name = "idx_traffic_incidents_location", columnList = "lat, lng"),
    @Index(name = "idx_traffic_incidents_level", columnList = "level"),
    @Index(name = "idx_traffic_incidents_status", columnList = "status"),
    @Index(name = "idx_traffic_incidents_created", columnList = "created_at")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class TrafficIncident {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @NotBlank(message = "Title is required")
    @Size(max = 255)
    @Column(nullable = false)
    private String title;
    
    @Column(columnDefinition = "TEXT")
    private String description;
    
    /**
     * Incident type: CAR, BIKE, ACCIDENT, JAM, SLOW, FAST
     * Determines which icon to display on frontend
     */
    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    private IncidentType type;
    
    /**
     * Severity level: LOW, MEDIUM, HIGH
     * Determines color coding and zone radius
     */
    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private SeverityLevel level;
    
    /**
     * Latitude (-90 to 90)
     */
    @NotNull
    @DecimalMin(value = "-90.0")
    @DecimalMax(value = "90.0")
    @Column(nullable = false)
    private Double lat;
    
    /**
     * Longitude (-180 to 180)
     */
    @NotNull
    @DecimalMin(value = "-180.0")
    @DecimalMax(value = "180.0")
    @Column(nullable = false)
    private Double lng;
    
    /**
     * Image data (can be URL or base64 encoded string)
     */
    @Column(columnDefinition = "TEXT")
    private String image;
    
    /**
     * Incident status
     */
    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private IncidentStatus status = IncidentStatus.PENDING;
    
    /**
     * Reporter name/identifier
     */
    @Column(length = 255)
    @Builder.Default
    private String reporter = "Anonymous";
    
    /**
     * Reference to affected traffic node (optional)
     */
    @Column(name = "affected_node_id")
    private Long affectedNodeId;
    
    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;
    
    @Column(name = "resolved_at")
    private LocalDateTime resolvedAt;
    
    // ==================== BUSINESS LOGIC ====================
    
    /**
     * Check if incident is high priority
     */
    public boolean isHighPriority() {
        return this.level == SeverityLevel.HIGH;
    }
    
    /**
     * Check if incident is validated
     */
    public boolean isValidated() {
        return this.status == IncidentStatus.VALIDATED;
    }
    
    /**
     * Check if incident has image
     */
    public boolean hasImage() {
        return this.image != null && !this.image.isEmpty();
    }
    
    /**
     * Mark incident as resolved
     */
    public void resolve() {
        this.status = IncidentStatus.RESOLVED;
        this.resolvedAt = LocalDateTime.now();
    }
    
    /**
     * Mark incident as validated
     */
    public void validate() {
        this.status = IncidentStatus.VALIDATED;
    }
    
    // ==================== ENUMS ====================
    
    /**
     * Incident Type - matches frontend expectations
     */
    public enum IncidentType {
        CAR,        // Car-related incident
        BIKE,       // Bike-related incident  
        ACCIDENT,   // Accident/crash
        JAM,        // Traffic jam
        SLOW,       // Slow moving traffic
        FAST        // Fast moving traffic (good)
    }
    
    /**
     * Severity Level - matches frontend expectations
     */
    public enum SeverityLevel {
        LOW,        // Minor issue
        MEDIUM,     // Moderate issue
        HIGH        // Severe issue
    }
    
    /**
     * Incident Status
     */
    public enum IncidentStatus {
        PENDING,    // Awaiting validation
        VALIDATED,  // Approved and displayed on map
        REJECTED,   // Rejected by moderator
        RESOLVED,   // Issue has been resolved
        ACTIVE      // Legacy status (same as VALIDATED)
    }
}
